pkg/cli/image/strategy/onerror.go

```diff
@@ -162,3 +154,185 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
+	var addSource bool
+	repo := imageRef.AsRepository().Exact()
+	for _, idms := range idmsList {
+		repoDigestMirrors := idms.Spec.ImageDigestMirrors
+		for _, rdm := range repoDigestMirrors {
+			var suffix string
+			var err error
+			rdmSourceRef, err := reference.Parse(rdm.Source)
+			if err != nil {
+				return nil, fmt.Errorf("invalid source %q: %w", rdm.Source, err)
+			}
+			// AsV2 in the right call is required to ensure we transform docker registry
+			// from docker.io to registry-1.docker.io
+			if imageRef.AsRepository().AsV2() != rdmSourceRef.AsRepository().AsV2() {
+				if !isSubrepo(repo, rdm.Source) {
+					continue
+				}
+				suffix = repo[len(rdm.Source):]
+			}
+			klog.V(5).Infof("%v ImageDigestMirrors source matches given image", imageRef.AsRepository().AsV2())
+			// check valid mirrorSourcePolicy
+			addSource, err = isAddSource(idmsList, rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			for _, m := range rdm.Mirrors {
+				mRef, err := reference.Parse(string(m) + suffix)
+				if err != nil {
+					return nil, fmt.Errorf("invalid mirror %q: %w", m, err)
+				}
+				imageSources = append(imageSources, mRef)
+				klog.V(5).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageDigestMirrorSet", m)
+			}
+		}
+	}
+
+	if addSource || len(imageSources) == 0 {
+		klog.V(5).Infof("%v ImageReference added to potential ImageSourcePrefixes from ImageDigestMirrorSet", imageRef.AsRepository().AsV2())
+		if addSourceAsLastAlternate {
+			imageSources = append(imageSources, imageRef.AsRepository().AsV2())
+		} else {
+			imageSources = append([]reference.DockerImageReference{imageRef.AsRepository().AsV2()}, imageSources...)
+		}
+	} else {
+		klog.V(5).Infof("%v ImageReference not added to potential ImageSourcePrefixes from ImageDigestMirrorSet", imageRef.AsRepository().AsV2())
+	}
+	uniqueMirrors := make([]reference.DockerImageReference, 0, len(imageSources))
+	uniqueMap := make(map[reference.DockerImageReference]bool)
+	for _, imageSourceMirror := range imageSources {
+		if _, ok := uniqueMap[imageSourceMirror]; !ok {
+			uniqueMap[imageSourceMirror] = true
+			uniqueMirrors = append(uniqueMirrors, imageSourceMirror)
+		}
+	}
+	klog.V(2).Infof("Found sources: %v for image: %v", uniqueMirrors, imageRef)
+	return uniqueMirrors, nil
+}
+
+func isAddSource(idmsList []apicfgv1.ImageDigestMirrorSet, source string) (bool, error) {
+	var found bool
+	var mirrorSourcePolicy apicfgv1.MirrorSourcePolicy
+	for _, idms := range idmsList {
+		for _, mirror := range idms.Spec.ImageDigestMirrors {
+			if mirror.Source != source {
+				continue
+			}
+			if !found && mirror.MirrorSourcePolicy != "" {
+				mirrorSourcePolicy = mirror.MirrorSourcePolicy
+				found = true
+				continue
+			}
+			if mirrorSourcePolicy != mirror.MirrorSourcePolicy {
+				return found, fmt.Errorf("ImageDigestMirrorSet can only contain one MirrorSourcePolicy for source %s", source)
+			}
```

Comment:
AllowContactingSource is the default value, empty mirror.MirrorSourcePolicy and AllowContactingSource are not equal but valid. eliminate that from the error return.

```suggestion
			if !found {
                if mirror.MirrorSourcePolicy != "" {
                    mirrorSourcePolicy = mirror.MirrorSourcePolicy
                } else {
                    mirrorSourcePolicy = apicfgv1.AllowContactingSource
                }
				found = true
				continue
			}
            if mirrorSourcePolicy == apicfgv1.AllowContactingSource && mirror.MirrorSourcePolicy == "" {
                continue
            }
            if mirrorSourcePolicy != mirror.MirrorSourcePolicy {
				return found, fmt.Errorf("ImageDigestMirrorSet can only contain one MirrorSourcePolicy for source %s", source)
			}
```

Reply:
@QiWang19 updated, PTAL?